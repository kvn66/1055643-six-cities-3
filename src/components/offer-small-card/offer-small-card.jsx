import React from "react";
import PropTypes from 'prop-types';
import {UNSELECTED_CARD_ID} from "../../const.js";
import {Link} from "react-router-dom";

const OfferSmallCard = (props) => {
  const {place, openDetail, setSelectedCard} = props;
  const {id, images, priceValue, priceText, name, type, isPremium} = place;
  const linkToDetail = `/offer/${id}`;

  const nameClickHandler = () => {
    openDetail(id);
  };

  const mouseEnterHandler = () => {
    setSelectedCard(id);
  };

  const mouseLeaveHandler = () => {
    setSelectedCard(UNSELECTED_CARD_ID);
  };

  return (
    <article onMouseEnter={mouseEnterHandler} onMouseLeave={mouseLeaveHandler} className="cities__place-card place-card">
      {isPremium &&
        <div className="place-card__mark">
          <span>Premium</span>
        </div>
      }
      <div className="cities__image-wrapper place-card__image-wrapper">
        <a href="#">
          <img className="place-card__image" src={images[0]} width="260" height="200" alt="Place image"/>
        </a>
      </div>
      <div className="place-card__info">
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{priceValue}</b>
            <span className="place-card__price-text">&#47;&nbsp;{priceText}</span>
          </div>
          <button
            className="place-card__bookmark-button place-card__bookmark-button--active button" type="button"
          >
            <svg className="place-card__bookmark-icon" width="18" height="19">
              <use xlinkHref="#icon-bookmark-active"/>
            </svg>
            <span className="visually-hidden">In bookmarks</span>
          </button>
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: `80%`}}/>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 onClick={nameClickHandler} className="place-card__name">
          <Link to={linkToDetail}>{name}</Link>
        </h2>
        <p className="place-card__type">{type}</p>
      </div>
    </article>
  );
};

OfferSmallCard.propTypes = {
  place: PropTypes.exact({
    id: PropTypes.number.isRequired,
    images: PropTypes.array.isRequired,
    priceValue: PropTypes.number.isRequired,
    priceText: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    type: PropTypes.string.isRequired,
    bedrooms: PropTypes.number.isRequired,
    adults: PropTypes.number.isRequired,
    rating: PropTypes.number.isRequired,
    inside: PropTypes.array.isRequired,
    isPremium: PropTypes.bool.isRequired,
    owner: PropTypes.exact({
      name: PropTypes.string.isRequired,
      avatar: PropTypes.string.isRequired,
      isSuper: PropTypes.bool.isRequired
    }).isRequired,
  }).isRequired,
  openDetail: PropTypes.func.isRequired,
  setSelectedCard: PropTypes.func.isRequired
};

export default OfferSmallCard;
