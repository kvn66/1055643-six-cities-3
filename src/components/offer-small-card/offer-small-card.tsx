import * as React from "react";
import {UNSELECTED_CARD_ID, HotelType} from "../../const";
import {AppRoute} from "../../const";
import {ActionCreator} from "../../reducers/card-select/card-select";
import {Operation as FavoriteOperation} from "../../reducers/favorites/favorites";
import {connect} from "react-redux";
import {CardType, ClassNameType} from "../../types";

type Props = {
  card: CardType;
  className: ClassNameType;
  setSelectedCardId: (id: number) => void;
  sendFavoriteStatus: (id: number) => void;
}

const OfferSmallCard: React.FunctionComponent<Props> = (props: Props) => {
  const {card, setSelectedCardId, sendFavoriteStatus, className} = props;
  const {id, previewImage, price, title, type, rating, isFavorite, isPremium} = card;
  const linkToDetail = `${AppRoute.OFFER}/${id}`;

  const mouseEnterHandler = () => {
    setSelectedCardId(id);
  };

  const mouseLeaveHandler = () => {
    setSelectedCardId(UNSELECTED_CARD_ID);
  };

  const mouseClickLinkHandler = () => {
    setSelectedCardId(UNSELECTED_CARD_ID);
  };

  const mouseClickFavoriteButtonHandler = (evt) => {
    evt.preventDefault();
    sendFavoriteStatus(id);
  };

  return (
    <article onMouseEnter={mouseEnterHandler} onMouseLeave={mouseLeaveHandler} className={`${className.ARTICLE} place-card`}>
      {isPremium &&
        <div className="place-card__mark">
          <span>Premium</span>
        </div>
      }
      <div className={`${className.IMAGE} place-card__image-wrapper`}>
        <a onClick={mouseClickLinkHandler} href={linkToDetail}>
          <img className="place-card__image" src={previewImage} width="260" height="200" alt="Place image"/>
        </a>
      </div>
      <div className="place-card__info">
        <div className="place-card__price-wrapper">
          <div className="place-card__price">
            <b className="place-card__price-value">&euro;{price}</b>
            <span className="place-card__price-text">&#47;&nbsp;night</span>
          </div>
          <button
            onClick={mouseClickFavoriteButtonHandler}
            className={`place-card__bookmark-button ${isFavorite ? `place-card__bookmark-button--active` : ``} button`}
            type="button"
          >
            <svg className="place-card__bookmark-icon" width="18" height="19">
              <use xlinkHref="#icon-bookmark"/>
            </svg>
            <span className="visually-hidden">In bookmarks</span>
          </button>
        </div>
        <div className="place-card__rating rating">
          <div className="place-card__stars rating__stars">
            <span style={{width: `${Math.round(rating) * 20}%`}}/>
            <span className="visually-hidden">Rating</span>
          </div>
        </div>
        <h2 className="place-card__name">
          <a onClick={mouseClickLinkHandler} href={linkToDetail}>{title}</a>
        </h2>
        <p className="place-card__type">{HotelType[type]}</p>
      </div>
    </article>
  );
};

const mapDispatchToProps = (dispatch) => {
  return {
    setSelectedCardId: (id) => dispatch(ActionCreator.setSelectedCardId(id)),
    sendFavoriteStatus: (id) => dispatch(FavoriteOperation.sendFavoriteStatus(id))
  };
};

export const MemoizedOfferSmallCard = connect(null, mapDispatchToProps)(React.memo(OfferSmallCard));

export default connect(null, mapDispatchToProps)(OfferSmallCard);
