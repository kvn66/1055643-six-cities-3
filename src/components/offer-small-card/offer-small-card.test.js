import React from 'react';
import renderer from 'react-test-renderer';
import {MemoizedOfferSmallCard} from "./offer-small-card.jsx";

const card = {
  id: 0,
  city: {
    location: {
      latitude: 52.38333,
      longitude: 4.9,
      zoom: 12,
    },
    name: `Amsterdam`,
  },
  bedrooms: 2,
  images: [`/img/apartment-01.jpg`, `/img/room.jpg`],
  description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
  goods: [`Wi-Fi`, `Washing machine`, `Towels`, `Heating`, `Coffee machine`, `Baby seat`, `Kitchen`, `Dishwasher`, `Cabel TV`, `Fridge`],
  host: {
    id: 0,
    name: `Angelina`,
    avatarUrl: `/img/avatar-angelina.jpg`,
    isPro: true
  },
  isFavorite: true,
  isPremium: true,
  location: {
    latitude: 52.3909553943508,
    longitude: 4.85309666406198,
    zoom: 12,
  },
  maxAdults: 3,
  previewImage: `/img/apartment-01.jpg`,
  price: 120,
  rating: 4.5,
  title: `Beautiful &amp; luxurious apartment at great location`,
  type: `apartment`,
};

it(`Render MemoizedOfferSmallCard`, () => {

  const tree = renderer
    .create(
        <MemoizedOfferSmallCard
          card={card}
          setSelectedCard={()=>{}}
          isDetail={false}
        />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
