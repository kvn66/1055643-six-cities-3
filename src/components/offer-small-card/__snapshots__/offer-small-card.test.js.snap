// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render OfferSmallCard 1`] = `
<div
  className="place-card__info"
>
  <div
    className="place-card__price-wrapper"
  >
    <div
      className="place-card__price"
    >
      <b
        className="place-card__price-value"
      >
        €
        120
      </b>
      <span
        className="place-card__price-text"
      >
        / 
        night
      </span>
    </div>
    <button
      className="place-card__bookmark-button place-card__bookmark-button--active button"
      type="button"
    >
      <svg
        className="place-card__bookmark-icon"
        height="19"
        width="18"
      >
        <use
          xlinkHref="#icon-bookmark-active"
        />
      </svg>
      <span
        className="visually-hidden"
      >
        In bookmarks
      </span>
    </button>
  </div>
  <div
    className="place-card__rating rating"
  >
    <div
      className="place-card__stars rating__stars"
    >
      <span
        style={
          Object {
            "width": "80%",
          }
        }
      />
      <span
        className="visually-hidden"
      >
        Rating
      </span>
    </div>
  </div>
  <h2
    className="place-card__name"
  >
    <a
      href="/offer/0"
    >
      Beautiful &amp; luxurious apartment at great location
    </a>
  </h2>
  <p
    className="place-card__type"
  >
    Apartment
  </p>
</div>
`;
